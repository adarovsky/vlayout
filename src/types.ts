import {Engine} from "./engine";

export class TypeDefinition {
    readonly  engine: Engine;
    readonly typeName: string;

    constructor(engine: Engine, typeName: string) {
        this.engine = engine;
        this.typeName = typeName;
    }

    toString(): string {
        return this.typeName;
    }
}

export interface Dictionary<T> {
    [Key: string]: T;
}

export class EnumDefinition extends TypeDefinition {
    private readonly values: Dictionary<any> = {};
    constructor(engine: Engine, typeName: string, values: Dictionary<any>) {
        super(engine, typeName);
        this.values = values;
    }
    valueFor(key: string): any {
        return this.values[key];
    }
}

export class Set extends TypeDefinition {
    readonly original: TypeDefinition;

    constructor(original: TypeDefinition) {
        super(original.engine, original.typeName + "*");
        this.original = original;
    }
}

export class Types {
    readonly engine: Engine;
    private _types: Dictionary<TypeDefinition> = {};

    private _straightValues(...values: string[]) {
        let r: Dictionary<string> = {};
        values.forEach(v => r[v] = v);
        return r;
    }

    constructor(engine: Engine) {
        this.engine = engine;

        ['String', 'Number', 'Bool', 'Font', 'Color', 'Image'].forEach( t => {
            this._types[t] = new TypeDefinition(engine, t);
        });
        ['String', 'Number'].forEach( t => {
            const s = new Set(this._types[t]);
            this._types[s.typeName] = s;
        });

        this.registerEnum(new EnumDefinition(engine, 'FontType', this._straightValues(
            'bold',
            'normal',
            'italic'
        )));

        this.registerEnum(new EnumDefinition(engine, 'TextAlignment', this._straightValues(
            'left',
            'right',
            'center',
            'top',
            'bottom',
            'middle'
        )));

        this.registerEnum(new EnumDefinition(engine, 'ImagePosition', this._straightValues(
            'left',
            'leftToText',
            'right',
            'rightToText',
            'top',
            'bottom'
        )));

        this.registerEnum(new EnumDefinition(engine, 'LayoutAlignment', this._straightValues(
            'fill',
            'center',
            'baseline',
            'top',
            'bottom',
            'leading',
            'trailing'
        )));

        this.registerEnum(new EnumDefinition(engine, 'ContentPolicy', this._straightValues(
            'fill',
            'aspectFit',
            'aspectFill',
            'center'
        )));

        this.registerEnum(new EnumDefinition(engine, 'SizePolicy', this._straightValues(
            'fixed',
            'stretched'
        )));

        this.registerEnum(new EnumDefinition(engine, 'GradientOrientation', this._straightValues(
            'topToBottom',
            'bottomToTop',
            'leftToRight',
            'rightToLeft'
        )));
    }

    registerEnum(e: EnumDefinition): void {
        const s = new Set(e);
        this._types[s.typeName] = s;
        this._types[e.typeName] = e;
    }

    type(named: string): TypeDefinition|null {
        return this._types[named] || null;
    }
}

export class ColorContainer {
    constructor(public readonly red: number,
                public readonly green: number,
                public readonly blue: number,
                public readonly alpha: number|null = null) {
    }

    toString(): string {
        if (this.alpha !== null) {
            return `rgba(${this.red}, ${this.green}, ${this.blue}, ${this.alpha})`;
        }
        else {
            return `rgb(${this.red}, ${this.green}, ${this.blue})`;
        }
    }
}

export class ImageContainer {
    constructor(public readonly src: string) {

    }
}

export class FontContainer {
    constructor(public readonly familyName: string|null, public readonly size: number|null, public readonly type: string|null) {
    }
}