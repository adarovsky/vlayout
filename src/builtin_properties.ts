import { TypeDefinition } from './types';
import { EMPTY, Observable } from 'rxjs';
import { Expression } from './expression';

export class BuiltinProperty extends Expression {
    constructor(readonly keyPath: string, readonly typeDefinition: TypeDefinition, readonly sink: Observable<any>) {
        super(0, 0);
    }

    instantiate(): this {
        const v = new (this.constructor as typeof BuiltinProperty)(this.keyPath, this.typeDefinition, this.sink);
        return v as this;
    }

    toString(): string {
        return this.keyPath;
    }
}
