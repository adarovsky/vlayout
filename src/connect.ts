import React, { createElement } from 'react';
import { combineLatest, Observable, of, Subscription } from 'rxjs';
import { fromPairs, toPairs, unzip, zip } from 'lodash';

export type ObservableMapObject<S = any> = {
    [K in keyof S]: Observable<S[K]>|S[K]
}

export function connect<P, S = {}>(WrappedComponent: React.FC<P> | React.ComponentClass<P, S>) {
    return class Connector extends React.PureComponent<ObservableMapObject<P>, P> {
        subscription = new Subscription();

        componentDidMount(): void {
            this.subscription.unsubscribe();
            const pairs = toPairs(this.props);
            const [keys, values] = unzip(pairs);
            this.subscription = combineLatest(values.map(v => v instanceof Observable ? v : of(v))).subscribe(s => {
                const pairs = zip(keys, s);
                const state = fromPairs(pairs);
                this.setState(state as P);
            });
        }

        componentWillUnmount(): void {
            this.subscription.unsubscribe();
        }

        render(): React.ReactElement<any, string | React.JSXElementConstructor<any>> | string | number | {} | React.ReactNodeArray | React.ReactPortal | boolean | null | undefined {
            if (this.state) {
                return createElement(WrappedComponent, this.state);
            } else {
                return null;
            }
        }
    }
}
