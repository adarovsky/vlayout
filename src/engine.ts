import { Inputs } from './inputs';
import {
    Dictionary,
    EnumDefinition,
    ListDefinition,
    ListDefinitionItem,
    TypeDefinition,
    Types,
} from './types';
import { EnumValue, Expression } from './expression';
import {
    Ceil,
    ColorAlpha,
    Floor,
    FontFamily,
    FontSized,
    FontTyped,
    FunctionImplementationI,
    Image,
    LocalizedNumber,
    LocalizedString,
    Max,
    Min,
    Round,
    ShortLocalizedNumber,
    StringPrefix,
    StringSuffix,
    SubString,
    ToCapitalCase,
    ToLower,
    ToUpper,
} from './builtin_functions';
import React from 'react';
import { ViewListReference, ViewReference } from './view_reference';
import { View } from './view';
import { ReactViewProps } from './react_views';
import { Button, TextField } from './primitives';
import { Observable } from 'rxjs';
import {
    ListClickHandler,
    ListEnterHandler,
    ListModelItem,
    ListTextChangeHandler,
} from './list';
import { isEqual } from 'lodash';
import invariant from 'ts-invariant';

export class Engine {
    readonly inputs = new Inputs(this);
    readonly types = new Types(this);
    readonly functions: FunctionImplementationI[];
    private readonly referencedViews: Dictionary<ViewReference> = {};
    private readonly buttons: Dictionary<() => Promise<void>> = {};
    private readonly textFields: Dictionary<{
        onChange: (s: string) => void;
        onEnter: () => void;
    }> = {};
    private readonly listButtons: Dictionary<ListClickHandler> = {};
    private readonly listTextFields: Dictionary<{
        onChange: ListTextChangeHandler;
        onEnter: ListEnterHandler;
    }> = {};
    private readonly referencedListViews: Dictionary<ViewListReference> = {};
    readonly valueSnapshot: {
        inputs: Dictionary<any>;
        properties: Dictionary<any>;
    } = {
        inputs: {},
        properties: {},
    };

    constructor(readonly debug = false, readonly verboseIds: string[] = []) {
        this.functions = [
            new LocalizedNumber(this),
            new ShortLocalizedNumber(this),
            new Image(this),
            new FontFamily(this),
            new FontTyped(this),
            new FontSized(this),
            new ColorAlpha(this),
            new StringPrefix(this),
            new StringSuffix(this),
            new SubString(this),
            new ToUpper(this),
            new ToLower(this),
            new ToCapitalCase(this),
            new Round(this),
            new Floor(this),
            new Ceil(this),
        ];
    }

    type(name: string): TypeDefinition | null {
        return this.types.type(name);
    }

    boolType(): TypeDefinition {
        return this.type('Bool')!;
    }

    numberType(): TypeDefinition {
        return this.type('Number')!;
    }

    colorType(): TypeDefinition {
        return this.type('Color')!;
    }

    stringType(): TypeDefinition {
        return this.type('String')!;
    }

    fontType(): TypeDefinition {
        return this.type('Font')!;
    }

    imageType(): TypeDefinition {
        return this.type('Image')!;
    }

    variableForKeyPath(keyPath: string): Expression | null {
        const kp = keyPath.split('.');
        if (kp.length === 2) {
            const enumType = this.type(kp[0]);
            if (enumType instanceof EnumDefinition) {
                const v = new EnumValue(kp[1], 0, 0);
                v.linkEnum(enumType);
                return v;
            }
        }

        return this.inputs.input(keyPath, 0, 0);
    }

    functionFor(
        name: string,
        parameters: TypeDefinition[]
    ): FunctionImplementationI {
        if (name === '@') {
            for (let p of parameters) {
                if (p !== this.stringType()) {
                    throw new Error(`localized string receives only strings`);
                }
            }

            return new LocalizedString(this);
        } else if (name === 'Min') {
            for (let p of parameters) {
                if (p !== this.numberType()) {
                    throw new Error(`Min accepts only numbers`);
                }
            }
            if (parameters.length == 0) {
                throw new Error(`Not enough arguments to Min`);
            }

            return new Min(this);
        } else if (name === 'Max') {
            for (let p of parameters) {
                if (p !== this.numberType()) {
                    throw new Error(`Max accepts only numbers`);
                }
            }

            if (parameters.length == 0) {
                throw new Error(`Not enough arguments to Max`);
            }

            return new Max(this);
        } else {
            for (let f of this.functions) {
                if (f.name === name && isEqual(f.parameterTypes, parameters)) {
                    return f;
                }
            }

            throw new Error(
                `function ${name}(${parameters
                    .map((p) => p.toString())
                    .join(', ')}) not found`
            );
        }
    }

    functionsLoose(
        name: string,
        parametersCount: number
    ): FunctionImplementationI[] {
        return this.functions.filter(
            (f) =>
                f.name === name && f.parameterTypes.length === parametersCount
        );
    }

    registerView(
        key: string,
        createComponent: (parent: View) => React.ReactElement<ReactViewProps>
    ) {
        invariant(
            typeof createComponent === 'function',
            `createComponent for ${key} is empty`
        );
        this.referencedViews[key] = new ViewReference(createComponent);
    }

    registerInput(
        name: string,
        type: TypeDefinition,
        sink: Observable<any>
    ): void {
        invariant(sink, `input sink for ${name} is empty`);
        this.inputs.registerInput(name, type, sink);
    }

    registerButton(key: string, onClick: () => Promise<void>) {
        this.buttons[key] = onClick;
    }

    registerTextField(
        key: string,
        onChange: (s: string) => void,
        sink: Observable<string>,
        onEnter: () => void = () => {}
    ) {
        this.textFields[key] = { onChange, onEnter };
        this.inputs.registerInput(key, this.stringType(), sink);
    }

    registerEnum(name: string, values: Dictionary<any>): void {
        this.types.registerEnum(new EnumDefinition(this, name, values));
    }

    registerStandardEnum<EnumType>(
        name: string,
        e: StandardEnum<EnumType>
    ): void {
        this.types.registerEnum(
            new EnumDefinition(this, name, materializeEnum(e))
        );
    }

    registerList(name: string, fields: Dictionary<ListDefinitionItem>): void {
        this.types.registerList(new ListDefinition(this, name, fields));
    }

    registerListButton(
        name: string,
        onClick: (item: ListModelItem) => Promise<void>
    ) {
        this.listButtons[name] = onClick;
    }

    registerListTextField(
        name: string,
        onChange: ListTextChangeHandler,
        onEnter: ListEnterHandler = () => {}
    ) {
        this.listTextFields[name] = { onChange, onEnter };
    }

    registerListView(
        name: string,
        createComponent: (
            parent: View,
            modelItem: Observable<ListModelItem>
        ) => React.ReactElement<ReactViewProps>
    ) {
        this.referencedListViews[name] = new ViewListReference(createComponent);
    }

    viewForKey(key: string): View | null {
        if (this.buttons[key]) return new Button(this.buttons[key]);
        else if (this.textFields[key])
            return new TextField(
                key,
                this.textFields[key].onChange,
                this.textFields[key].onEnter
            );
        else if (this.referencedViews[key])
            return new ViewReference(this.referencedViews[key].createComponent);
        else return null;
    }

    listButtonForKey(key: string): ListClickHandler | null {
        return this.listButtons[key] || null;
    }

    listTextFieldForKey(
        key: string
    ): { onChange: ListTextChangeHandler; onEnter: ListEnterHandler } | null {
        return this.listTextFields[key] || null;
    }

    listViewForKey(key: string): ViewListReference | null {
        const view = this.referencedListViews[key];
        if (view) {
            return new ViewListReference(view.createComponent);
        }

        return null;
    }

    logInputValue(name: string, value: any): void {
        this.valueSnapshot.inputs[name] = value;
    }

    beginUpdates() {
        this.inputs.beginUpdates();
    }

    endUpdates() {
        this.inputs.endUpdates();
    }
}

export type StandardEnum<T> = {
    [id: string]: T | string;
    [nu: number]: string;
};

function camelize(str: string) {
    return str
        .replace(/(?:^\w|[A-Z]|\b\w|-|\s+)/g, function (match, index) {
            if (+match === 0) return ''; // or if (/\s+/.test(match)) for white spaces
            return index === 0 ? match.toLowerCase() : match.toUpperCase();
        })
        .replace('-', '');
}

function materializeEnum<EnumType>(e: StandardEnum<EnumType>) {
    return Object.keys(e)
        .filter((x) => isNaN(+x))
        .reduce((prev, cur) => ({ ...prev, [camelize(cur)]: e[cur] }), {});
}
