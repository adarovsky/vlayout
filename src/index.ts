export {Engine} from "./engine";
export {Layout} from "./layout";
export {ReactView, ReactViewState} from "./react_views";
import './vlayout.css';
