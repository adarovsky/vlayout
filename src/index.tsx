export {
    resizeObserver,
    pauseObserving,
    resumeObserving,
} from './resize_sensor';
export { Engine } from './engine';
export { Layout } from './layout';
import './vlayout.css';

export { ReactViewReference as ReactView } from './react_views';
export type { ReactViewProps, ReactViewState } from './react_views';
export { View } from './view';
export { ColorContainer as Color, ImageContainer as Image } from './types';
export type {
    SimpleListDefinitionItem,
    ListDefinitionItem,
    ComplexDefinitionItem,
} from './types';
export type { ListModelItem } from './list';
export { LayoutComponent } from './layout_component';
export { connect } from './connect';
export type { ObservableMapObject } from './connect';
