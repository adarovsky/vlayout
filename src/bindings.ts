import { Engine } from 'engine';
import { View } from 'view';
import { Button, TextField } from 'primitives';

export class Bindings {
    constructor(public readonly engine: Engine) {

    }

    registerSlot(name: string, type: string): void {
        if (type === 'listButton' && this.engine.listButtonForKey(name))
            return;

        if (type === 'listTextField' && this.engine.listTextFieldForKey(name))
            return;

        if (type === 'listView' && this.engine.listViewForKey(name))
            return;

        const v = this.viewForKey(name);
        if (!v) {
            throw new Error(`view ${name} is not registered`);
        }

        if (type === 'button' && !(v instanceof Button)) {
            throw new Error(`view ${name} is not a button, but ${v.viewType()}`);
        }

        if (type === 'textField' && !(v instanceof TextField)) {
            throw new Error(`view ${name} is not a text field, but ${v.viewType()}`);
        }
    }

    viewForKey(key: string): View | null {
        return this.engine.viewForKey(key);
    }

    toString(): string {
        return 'bindings {\n}\n';
    }
}
