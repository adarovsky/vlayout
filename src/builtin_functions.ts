import {Engine} from "./engine";
import {ColorContainer, FontContainer, ImageContainer, TypeDefinition} from "./types";
import {combineLatest, Observable} from "rxjs";
import {map} from "rxjs/operators";

export interface FunctionImplementationI {
    returnType: TypeDefinition;
    name: string;
    parameterTypes: TypeDefinition[];

    sink(parameters: Observable<any>[]): Observable<any>;
}

export class FunctionImplementation {
    constructor(public readonly engine: Engine) {
    }
}

export class LocalizedNumber extends FunctionImplementation implements FunctionImplementationI {
    name: string;
    parameterTypes: TypeDefinition[];
    returnType: TypeDefinition;

    constructor(engine: Engine) {
        super(engine);

        this.name = 'String';
        this.parameterTypes = [engine.numberType()];
        this.returnType = engine.stringType();
    }

    sink(parameters: Observable<any>[]): Observable<any> {
        return parameters[0].pipe(
            map( x => `${x}`)
        );
    }
}

export class ShortLocalizedNumber extends FunctionImplementation implements FunctionImplementationI {
    name: string;
    parameterTypes: TypeDefinition[];
    returnType: TypeDefinition;

    constructor(engine: Engine) {
        super(engine);

        this.name = 'ShortString';
        this.parameterTypes = [engine.numberType()];
        this.returnType = engine.stringType();
    }

    sink(parameters: Observable<any>[]): Observable<any> {
        return parameters[0].pipe(
            map( x => {
                if (x > 1000000) {
                    return `${Math.round(x / 100000) / 10} M`
                }
                else if (x > 1000) {
                    return `${Math.round(x / 100) / 10} K`
                }
                return `${x}`;
            })
        );
    }
}

export class LocalizedString extends FunctionImplementation implements FunctionImplementationI {
    name: string;
    parameterTypes: TypeDefinition[];
    returnType: TypeDefinition;

    constructor(engine: Engine) {
        super(engine);

        this.name = 'String';
        this.parameterTypes = [engine.stringType()];
        this.returnType = engine.stringType();
    }

    sink(parameters: Observable<any>[]): Observable<any> {
        return combineLatest(parameters as Observable<string>[]).pipe(
            map(params => {
                let key = params[0];
                for (let i = params.length - 1; i >= 1; --i) {
                    key = key.replace(`$${i}`, params[i]);
                }

                return key;
            })
        );
    }
}

export class Image extends FunctionImplementation implements FunctionImplementationI {
    name: string;
    parameterTypes: TypeDefinition[];
    returnType: TypeDefinition;

    constructor(engine: Engine) {
        super(engine);

        this.name = 'Image';
        this.parameterTypes = [engine.stringType()];
        this.returnType = engine.imageType();
    }

    sink(parameters: Observable<any>[]): Observable<any> {
        return parameters[0].pipe(
            map(x => new ImageContainer(x))
        );
    }
}

export class FontFamily extends FunctionImplementation implements FunctionImplementationI {
    name: string;
    parameterTypes: TypeDefinition[];
    returnType: TypeDefinition;

    constructor(engine: Engine) {
        super(engine);

        this.name = 'Font';
        this.parameterTypes = [engine.stringType(), engine.numberType()];
        this.returnType = engine.fontType();
    }

    sink(parameters: Observable<any>[]): Observable<any> {
        return combineLatest(parameters as Observable<any>[]).pipe(
            map(params => new FontContainer(params[0], params[1], null))
        );
    }
}

export class FontTyped extends FunctionImplementation implements FunctionImplementationI {
    name: string;
    parameterTypes: TypeDefinition[];
    returnType: TypeDefinition;

    constructor(engine: Engine) {
        super(engine);

        this.name = 'Font';
        this.parameterTypes = [engine.type('FontType')!, engine.numberType()];
        this.returnType = engine.fontType();
    }

    sink(parameters: Observable<any>[]): Observable<any> {
        return combineLatest(parameters as Observable<any>[]).pipe(
            map(params => new FontContainer(null, params[1], params[0]))
        );
    }
}

export class FontSized extends FunctionImplementation implements FunctionImplementationI {
    name: string;
    parameterTypes: TypeDefinition[];
    returnType: TypeDefinition;

    constructor(engine: Engine) {
        super(engine);

        this.name = 'Font';
        this.parameterTypes = [engine.numberType()];
        this.returnType = engine.fontType();
    }

    sink(parameters: Observable<any>[]): Observable<any> {
        return parameters[0].pipe(
            map(params => new FontContainer(null, params, null))
        );
    }
}

export class ColorAlpha extends FunctionImplementation implements FunctionImplementationI {
    name: string;
    parameterTypes: TypeDefinition[];
    returnType: TypeDefinition;

    constructor(engine: Engine) {
        super(engine);

        this.name = 'Color';
        this.parameterTypes = [engine.colorType(), engine.numberType()];
        this.returnType = engine.colorType();
    }

    sink(parameters: Observable<any>[]): Observable<any> {
        return combineLatest(parameters as Observable<any>[]).pipe(
            map(params => {
                const original = params[0] as ColorContainer;
                const alpha = params[1] as number;
                return new ColorContainer(original.red, original.green, original.blue, Math.max(0, Math.min(alpha, 1)));
            })
        );
    }
}