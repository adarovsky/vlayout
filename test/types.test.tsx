import { mount, shallow } from 'enzyme';
import { Engine, Layout } from '../src';
import React from 'react';
import { Subject } from 'rxjs';
import { ImageContainer } from '../src/types';

let engine = new Engine();

beforeEach(() => {
    engine = new Engine({ debug: true });
});

beforeAll(() => {
    // Create a spy on console (console.log in this case) and provide some mocked implementation
    // In mocking global objects it's usually better than simple `jest.fn()`
    // because you can `unmock` it in clean way doing `mockRestore`
    jest.spyOn(console, 'error').mockImplementation(() => {});
});
afterAll(() => {
    // Restore mock after all tests are done, so it won't affect other test suites
    // @ts-ignore
    console.error.mockRestore();
});
afterEach(() => {
    // Clear mock (all calls etc) after each test.
    // It's needed when you're using console somewhere in the tests so you have clean mock each time
    // @ts-ignore
    console.error.mockClear();
});

describe("types", () => {
    it("should throw if invalid value is sent to input", async function () {
        const subj = new Subject<any>();
        engine.registerInput('testInput', engine.stringType(), subj);
        const wrapper = mount(
            <Layout
                engine={engine}
                content={`
                 inputs {
                    testInput: String
                 }
                 layout {
                     layer {
                         label {
                             id: "label1"
                             center { x: 0.5 y: 0.5 }
                             
                             text: testInput                 
                         }
                     }
                 }`}
            />
        );

        const node = wrapper.find(".vlayout_label");
        subj.next(123);

        expect(console.error).toHaveBeenCalled();
        expect(node.getDOMNode()).toMatchSnapshot();

        subj.next('abcde');
        expect(node.getDOMNode()).toMatchSnapshot();
    });

    it("should throw if invalid enum value is sent to input", async function () {
        const subj = new Subject<any>();
        engine.registerEnum('TestEnum', {'a': 1, 'b': 2});
        engine.registerInput('testInput', engine.type('TestEnum')!, subj);
        const wrapper = mount(
            <Layout
                engine={engine}
                content={`
                 inputs {
                    testInput: TestEnum
                 }
                 layout {
                     layer {
                         label {
                             id: "label1"
                             center { x: 0.5 y: 0.5 }
                             
                             text: testInput == .a ? 'a' : 'other'                 
                         }
                     }
                 }`}
            />
        );

        const node = wrapper.find(".vlayout_label");
        subj.next(null);
        expect(console.error).toHaveBeenCalledTimes(1);

        subj.next(10);
        expect(console.error).toHaveBeenCalledTimes(2);

        expect(node.getDOMNode()).toMatchSnapshot();

        subj.next(1);
        expect(console.error).toHaveBeenCalledTimes(2);
        expect(node.getDOMNode()).toMatchSnapshot();
    });

    it("should allow null to number inputs", async function () {
        const subj = new Subject<any>();
        engine.registerInput('testInput', engine.numberType(), subj);
        const wrapper = mount(
            <Layout
                engine={engine}
                content={`
                 inputs {
                    testInput: Number
                 }
                 layout {
                     layer {
                         label {
                             id: "label1"
                             center { x: 0.5 y: 0.5 }
                             
                             text: testInput == nil ? "null" : String(testInput)                 
                         }
                     }
                 }`}
            />
        );

        const node = wrapper.find(".vlayout_label");
        subj.next(null);

        expect(console.error).toHaveBeenCalledTimes(0);
        expect(node.getDOMNode()).toMatchSnapshot();

        subj.next('abcde');
        expect(console.error).toHaveBeenCalledTimes(1);

        subj.next(10);
        expect(console.error).toHaveBeenCalledTimes(1);

        expect(node.getDOMNode()).toMatchSnapshot();
    });

    it("should warn about missing value", async function () {
        jest.useFakeTimers();
        const subj = new Subject<any>();
        engine.registerInput('testInput', engine.numberType(), subj);
        const wrapper = mount(
            <Layout
                engine={engine}
                content={`
                 inputs {
                    testInput: Number
                 }
                 layout {
                     layer {
                         label {
                             id: "label1"
                             center { x: 0.5 y: 0.5 }
                             
                             text: testInput == nil ? "null" : String(testInput)                 
                         }
                     }
                 }`}
            />
        );

        expect(() => jest.advanceTimersByTime(2000)).toThrow(/inconsistency for input testInput: no value came in/);
    });

    it("should accept nil values in functions", async function () {
        const subj = new Subject<any>();
        engine.registerInput('testInput', engine.numberType(), subj);
        const wrapper = shallow(
            <Layout
                engine={engine}
                content={`
                 inputs {
                    testInput: Number
                 }
                 functions {
                    testFunction(index:Number) => testInput <= 1 ? nil : index % 2 == 0 ? 10 : nil
                 }
                 layout {
                     layer {
                         label {
                             id: "label1"
                             center { x: 0.5 y: 0.5 }
                             
                             text: String(testInput)                 
                         }
                     }
                 }`}
            />
        );

        expect(() => wrapper).not.toThrow()
    });

    it('Image should support prefix for relative urls', async function () {
        engine = new Engine({ debug: true, prefix: '/prefix' });
        const i = new ImageContainer('/some/path.png', engine);
        expect(i.src).toBe('/prefix/some/path.png');
    });

    it('Image should include slash if missing for relative path', async function () {
        engine = new Engine({ debug: true, prefix: '/prefix' });
        const i = new ImageContainer('some/path.png', engine);
        expect(i.src).toBe('/prefix/some/path.png');
    });

    it('Image should include single slash if missing for relative path', async function () {
        engine = new Engine({ debug: true, prefix: '/' });
        const i = new ImageContainer('some/path.png', engine);
        expect(i.src).toBe('/some/path.png');
    });

    it('Image should not duplicate slashes if prefix ends with slash', async function () {
        engine = new Engine({ debug: true, prefix: '/prefix/' });
        const i = new ImageContainer('some/path.png', engine);
        expect(i.src).toBe('/prefix/some/path.png');
    });

    it('Image should not duplicate slashes if prefix ends with slash and url starts with slash', async function () {
        engine = new Engine({ debug: true, prefix: '/prefix/' });
        const i = new ImageContainer('/some/path.png', engine);
        expect(i.src).toBe('/prefix/some/path.png');
    });

    it('Image should not insert slash for relative path without slash', async function () {
        const i = new ImageContainer('some/path.png', engine);
        expect(i.src).toBe('some/path.png');
    });

    it('Image should keep full url as is', async function () {
        engine = new Engine({ debug: true, prefix: '/prefix' });
        const i = new ImageContainer('http://server/some/path.png', engine);
        expect(i.src).toBe('http://server/some/path.png');
    });
});

