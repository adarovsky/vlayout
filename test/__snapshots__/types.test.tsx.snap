// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`types should allow null to number inputs 1`] = `
<div
  class="vlayout_label"
  id="label1"
  style="position: absolute; left: 50%; transform: translateX(-50%) translateY(-50%); z-index: 1; top: 50%; text-align: start; overflow: hidden; text-overflow: ellipsis;"
>
  <span>
    null
    <br />
  </span>
  <div
    class="vlayout_label_shadow"
    style="position: absolute; white-space: pre; opacity: 0; transform: translateX(-50%) translateY(-50%); z-index: 1; text-align: start; overflow: hidden; text-overflow: ellipsis;"
  >
    <span>
      null
      <br />
    </span>
  </div>
</div>
`;

exports[`types should allow null to number inputs 2`] = `
<div
  class="vlayout_label"
  id="label1"
  style="position: absolute; left: 50%; transform: translateX(-50%) translateY(-50%); z-index: 1; top: 50%; text-align: start; overflow: hidden; text-overflow: ellipsis;"
>
  <span>
    10
    <br />
  </span>
  <div
    class="vlayout_label_shadow"
    style="position: absolute; white-space: pre; opacity: 0; transform: translateX(-50%) translateY(-50%); z-index: 1; text-align: start; overflow: hidden; text-overflow: ellipsis;"
  >
    <span>
      10
      <br />
    </span>
  </div>
</div>
`;

exports[`types should throw if invalid enum value is sent to input 1`] = `
<div
  class="vlayout_label"
  id="label1"
  style="position: absolute; left: 50%; transform: translateX(-50%) translateY(-50%); z-index: 1; top: 50%; text-align: start; overflow: hidden; text-overflow: ellipsis;"
>
  <span>
    
    <br />
  </span>
  <div
    class="vlayout_label_shadow"
    style="position: absolute; white-space: pre; opacity: 0; transform: translateX(-50%) translateY(-50%); z-index: 1; text-align: start; overflow: hidden; text-overflow: ellipsis;"
  >
    <span>
      
      <br />
    </span>
  </div>
</div>
`;

exports[`types should throw if invalid enum value is sent to input 2`] = `
<div
  class="vlayout_label"
  id="label1"
  style="position: absolute; left: 50%; transform: translateX(-50%) translateY(-50%); z-index: 1; top: 50%; text-align: start; overflow: hidden; text-overflow: ellipsis;"
>
  <span>
    a
    <br />
  </span>
  <div
    class="vlayout_label_shadow"
    style="position: absolute; white-space: pre; opacity: 0; transform: translateX(-50%) translateY(-50%); z-index: 1; text-align: start; overflow: hidden; text-overflow: ellipsis;"
  >
    <span>
      a
      <br />
    </span>
  </div>
</div>
`;

exports[`types should throw if invalid value is sent to input 1`] = `
<div
  class="vlayout_label"
  id="label1"
  style="position: absolute; left: 50%; transform: translateX(-50%) translateY(-50%); z-index: 1; top: 50%; text-align: start; overflow: hidden; text-overflow: ellipsis;"
>
  <span>
    
    <br />
  </span>
  <div
    class="vlayout_label_shadow"
    style="position: absolute; white-space: pre; opacity: 0; transform: translateX(-50%) translateY(-50%); z-index: 1; text-align: start; overflow: hidden; text-overflow: ellipsis;"
  >
    <span>
      
      <br />
    </span>
  </div>
</div>
`;

exports[`types should throw if invalid value is sent to input 2`] = `
<div
  class="vlayout_label"
  id="label1"
  style="position: absolute; left: 50%; transform: translateX(-50%) translateY(-50%); z-index: 1; top: 50%; text-align: start; overflow: hidden; text-overflow: ellipsis;"
>
  <span>
    abcde
    <br />
  </span>
  <div
    class="vlayout_label_shadow"
    style="position: absolute; white-space: pre; opacity: 0; transform: translateX(-50%) translateY(-50%); z-index: 1; text-align: start; overflow: hidden; text-overflow: ellipsis;"
  >
    <span>
      abcde
      <br />
    </span>
  </div>
</div>
`;
